Reference implementations for rot90, roll, atleast_1d,2d,3d

IvanYashchuk opened the pull request:
This PR adds the following references:

- `rot90`
- `roll`
- `atleast_1d`
- `atleast_2d`
- `atleast_3d`

IvanYashchuk committed:
Add rot90 ref

IvanYashchuk committed:
Update sample inputs of rot90

IvanYashchuk committed:
Add roll ref

IvanYashchuk committed:
Add rot90 and roll to __all__

IvanYashchuk committed:
Add atleast_1d,2d,3d

IvanYashchuk committed:
Enable OpInfo-based tests for rot90

IvanYashchuk committed:
Enable OpInfo-based tests for roll

IvanYashchuk committed:
Enable OpInfo-based tests for atleast_1d,2d,3d

IvanYashchuk committed:
Fix typo

IvanYashchuk committed:
Specify return type of atleast functions as Union[TensorLikeType, Tuple[TensorLikeType]]

IvanYashchuk committed:
Change DimsType -> DimsSequenceType

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        raise ValueError(f"expected total rotation dims == 2, but got dims = {len(dims)}")
+    if a.ndim < 2:
+        raise ValueError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and (dims[0] - dims[1]).__abs__() != a.ndim):
```
Regular Python abs is probably more readable than the dunder here?

mruberry replied:
I know and that's lame but we should get consistent about calling `builtins.abs` directly (and more generally calling `builtins.foo` whenever we clobber a Python name)

IvanYashchuk replied:
Thanks! Let's use `builtins.abs`.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
 
+def rot90(a: TensorLikeType, k: int = 1, dims: DimsSequenceType = (0, 1)) -> TensorLikeType:
+    """Reference implementation of :func:`torch.rot90`."""
+    if len(dims) != 2:
```
I think this needs to canonicalize the dims upfront with utils.canonicalize_dims (which wraps dims specified as negative numbers)

IvanYashchuk replied:
Initially I thought that `canonicalize_dims` is not needed here because `dims` is passed to `flip` and `transpose` which themselves call `canonicalize_dims` taking care of negative numbers. But then I see now that the "out of range" checks are unnecessarily duplicated in this function.

mruberry replied:
Eventually we should investigate an error context so even if an error is thrown from another operation it cites the user-facing entrypoint as expected.

However, we don't have an error context today, so I would suggest duplicating error checks for the moment both for message quality and for code simplicity -- because it can sometimes be tricky to reason about what other operations will or will not validate

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        raise ValueError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and (dims[0] - dims[1]).__abs__() != a.ndim):
+        raise ValueError(f"expected rotation dims to be different, but got dim0 = {dims[0]} and dim1 = {dims[1]}")
+    if not (dims[0] < a.ndim and dims[0] >= -a.ndim):
```
dim canonicalization should take care of these checks, too

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        first_dim_rolled = roll(a, shifts[0], dims[0])
+        return roll(first_dim_rolled, tail_shifts, tail_dims)
+
+    if a.numel == 0:
```
Comment for this sho-circuit -- should this be sooner?

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+    start = (size - shifts[0]) % size
+    t0 = narrow(a, dim, start, size - start)
+    t1 = narrow(a, dim, 0, start)
+    return cat((t0, t1), dim)
```
Wow this is a crazy function

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6718,11 +6718,9 @@ def sample_inputs_roll(op_info, device, dtype, requires_grad=False, **kwargs):
-            (1, (1, 2),),
-            (1, (1, -1),),
-            ())
+    args = itertools.product(range(-5, 6), [(0, 1), (1, 2), (1, -1)])
```
Nice

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19230,6 +19240,14 @@ def __init__(
         torch_opinfo_name="reshape",
     ),
+    PythonRefInfo(
+        "_refs.rot90",
```
This looks really good -- are the sample inputs for rot90 and/or roll sufficient, or should we use this opportunity to extend them with some reference inputs, too, to ensure all code paths and edge cases are validated? 

Also, although it is more work, error inputs for rot90 and roll would be great to verify their many error checks, but your call whether it's worth going there at this time

mruberry commented:
Overall looks good, @IvanYashchuk -- just a few questions and suggestions inline, and we should wait for the tests. Looking forward to hearing your thoughts on adding references!

IvanYashchuk committed:
Sort rot90, roll

IvanYashchuk committed:
Replace __abs__ -> builtins.abs

IvanYashchuk committed:
Use canonicalize_dims in rot90

IvanYashchuk committed:
Use canonicalize_dims in roll

IvanYashchuk committed:
Add a comment for len_dims == 0 and len_shifts == 1 case

IvanYashchuk committed:
Add 0 numel sample input

IvanYashchuk committed:
Roll before rot90

IvanYashchuk committed:
Add a comment to roll

IvanYashchuk committed:
Add error_inputs_roll

IvanYashchuk committed:
Add one more ErrorInput to roll

IvanYashchuk committed:
Add out of range/bounds error input

IvanYashchuk committed:
Add error_inputs_rot90

IvanYashchuk committed:
:Merge remote-tracking branch 'upstream/viable/strict' into refs-rot90

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into refs-rot90

IvanYashchuk committed:
Formatting

IvanYashchuk committed:
Try fixing mypy errors

IvanYashchuk committed:
Black formatting

IvanYashchuk committed:
Fix MyPy error for atleast functions

IvanYashchuk committed:
Fix mypy errors in rot90

IvanYashchuk committed:
Fix _unsqueeze_atleast

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into refs-rot90

IvanYashchuk committed:
Remove unrelated changes

IvanYashchuk commented:
I added error inputs for `roll` and `rot90` and addressed review feedback. It seems the XLA failure is unrelated.

@mruberry, could you please take another look?

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        if len_shifts == 0:
+            raise ValueError("`shifts` required")
+        if len_dims == 0 and len_shifts == 1:
+            return clone(view(roll(flatten(a), shifts, 0), a.shape))
```
If clone is just to pacify the tests, maybe we should disable them (we disable them for reduction ops e.g.)

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1784,6 +1830,80 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        )
+    if a.ndim < 2:
+        raise RuntimeError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and builtins.abs(dims[0] - dims[1]) != a.ndim):
```
second part of conditional cannot happen after canonicalization?

IvanYashchuk replied:
Right. I will remove it.

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -1784,6 +1830,80 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
+        )
+    k = k % 4  # Rotation direction is from the second towards the first axis for k < 0
+    if k == 1:
+        return clone(transpose(flip(a, (dims[1],)), dims[0], dims[1]))
```
are clones here again to pacify view tests?

IvanYashchuk committed:
Don't validate view consistency for rot90 and roll

IvanYashchuk committed:
Remove the second part of the conditional dims[0] == dims[1]

mruberry reviewed (approved):
Test failures are unrelated -- I think this is good to go!

Thanks @IvanYashchuk -- curious to hear your thoughts about where we should go next

ngimel commented:
@pytorchbot merge this

malfet commented:
@pytorchbot revert this broke XLA testing on master (and it were reported on PR as well), i.e. see https://github.com/pytorch/pytorch/pull/78080 (PR CI) and https://hud.pytorch.org/pytorch/pytorch/commit/96c134854d4dbc418cdc0ec82959476ddac8068e (master CI)

mruberry commented:
@ezyang are we auto-registering something internally that could have caused this to break XLA? I don't think we're running any primTorch tests on XLA

malfet commented:
@pytorchbot revert this as it broke XLA on trunk (see https://github.com/pytorch/pytorch/runs/6678429656?check_suite_focus=true ) and the same pattern were observable on PR CI https://github.com/pytorch/pytorch/runs/6672733779?check_suite_focus=true

JackCaoG commented:
Thanks, I just saw the XLA test failed at `BuildRoll`. Feel free to open a ticket on xla side, I will take a look why test crashed.

IvanYashchuk committed:
Add onlyNativeDeviceTypes to roll's OpInfo because XLA is failing

IvanYashchuk commented:
@pytorchbot merge this

backwards_compat failure is unrelated:
```
Broken ops: [
	aten::special_laguerre_polynomial_l(Tensor x, Tensor n) -> Tensor
	aten::special_laguerre_polynomial_l.out(Tensor x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.n_scalar_out(Tensor x, Scalar n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.x_scalar(Scalar x, Tensor n) -> Tensor
	aten::special_laguerre_polynomial_l.x_scalar_out(Scalar x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.n_scalar(Tensor x, Scalar n) -> Tensor
	aten::special_hermite_polynomial_he(Tensor x, Tensor n) -> Tensor
	aten::special_hermite_polynomial_he.out(Tensor x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.n_scalar_out(Tensor x, Scalar n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.x_scalar(Scalar x, Tensor n) -> Tensor
	aten::special_hermite_polynomial_he.x_scalar_out(Scalar x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.n_scalar(Tensor x, Scalar n) -> Tensor
	aten::_linalg_svd(Tensor A, bool full_matrices=False, bool compute_uv=True, *, str? driver=None) -> (Tensor U, Tensor S, Tensor Vh)
	aten::_linalg_svd.U(Tensor A, bool full_matrices=False, bool compute_uv=True, *, str? driver=None, Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh) -> (Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh)
	aten::linalg_svd(Tensor A, bool full_matrices=True, *, str? driver=None) -> (Tensor U, Tensor S, Tensor Vh)
	aten::linalg_svd.U(Tensor A, bool full_matrices=True, *, str? driver=None, Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh) -> (Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh)
	aten::linalg_svdvals(Tensor A, *, str? driver=None) -> Tensor
	aten::linalg_svdvals.out(Tensor A, *, str? driver=None, Tensor(a!) out) -> Tensor(a!)
]
```

IvanYashchuk commented:
@pytorchbot merge
